presto-client/src/main/java/io/prestosql/client/auth/external/Tokens.java

```diff
@@ -0,0 +1,128 @@
+import static java.net.HttpURLConnection.HTTP_OK;
+import static java.net.HttpURLConnection.HTTP_UNAVAILABLE;
+import static java.util.Objects.requireNonNull;
+
+class Tokens
+{
+    public static final int HTTP_FAILED_DEPENDENCY = 424;
+    private final OkHttpClient client;
+    private final ObjectMapper mapper = new ObjectMapper();
+    private Duration timeout;
+
+    Tokens(OkHttpClient client, Duration timeout)
+    {
+        this.client = requireNonNull(client, "client is null");
+        this.timeout = requireNonNull(timeout, "timeout is null");
+    }
+
+    public TokenPoll pollForToken(String tokenUrl)
+            throws TokenPollException
+    {
+        try {
+            return Failsafe.with(RetryPolicies.<TokenPoll>retryUntil(timeout)
+                    .abortIf(TokenPoll::nextTokenUriExists)
+                    .handleResultIf(TokenPoll::isNotAvailableYet)
+                    .handleIf(tx -> false))
+                    .get(context -> {
+                        Call call = Optional.<TokenPoll>ofNullable(context.getLastResult())
+                                .flatMap(TokenPoll::getNextTokenUri)
+                                .map(this::createCallFor)
+                                .orElseGet(() -> createCallFor(tokenUrl));
+                        return handleResponse(call);
+                    });
+        }
+        catch (FailsafeException e) {
+            // OkHttp throws this after clearing the interrupt status
+            // TODO: remove after updating to Okio 1.15.0+
+            Throwable cause = e.getCause();
+            if (InterruptedIOException.class.equals(cause.getClass()) && "thread interrupted".equals(cause.getMessage())) {
+                Thread.currentThread().interrupt();
+            }
+            throw new TokenPollException(cause);
+        }
+    }
+
+    private Call createCallFor(String tokenUrl)
+    {
+        return client.newCall(new Request.Builder()
+                .url(tokenUrl)
+                .get()
+                .build());
+    }
+
+    private TokenPoll handleResponse(Call call)
+            throws IOException
+    {
+        try (Response response = call.execute()) {
+            switch (response.code()) {
+                case HTTP_OK:
+                    try (ResponseBody body = response.body()) {
+                        TokenPollDTO pollDto = mapper.readValue(body.string(), TokenPollDTO.class);
+                        return pollDto.toTokenPoll();
+                    }
+                case HTTP_UNAVAILABLE:
+                    try (ResponseBody body = response.body()) {
+                        throw new IOException(format("Token poll failed with message: %s", body.string()));
+                    }
+                default:
+                    return TokenPoll.failed(format("Unknown response code \"%s\", retrieved from token poll", response.code()));
+            }
+        }
+    }
+
+    private static class TokenPollDTO
```

Comment:
Why not to have `TokenPoll` to be DTO friendly? It would make it simpler, IMO `TokenPoll` looks like a DTO already.